### ISSUE TITLE
Include function name in report message if possible / relevant

### ISSUE THREAD

**Issue by Moeriki** [eslint/eslint#7260]
I have these two warning about the same function in my code.
- **complexity** Function 'formatLog' has a complexity of 12.
- **max-statements** This function has too many statements (25). Maximum allowed is 16.

Notice that the latter says `This function`, where the former says `Function 'formatLog'`. This named notation I find really helpful.

I did a quick search in the rules and found a few more mentioning a function that could potentially be named.

---
## What rule do you want to change?
- [x] array-callback-return
- [x] complexity
- [x] consistent-return
- [x] func-names
- [x] max-params
- [x] max-statements
- [x] no-empty-functions
- [x] strict

I excluded

**1.** rules where the function should not be the focus of the feedback message. Eg. `no-dupe-args` gives `Duplicate param '{{name}}'`. Since we already name the param, I don't find it more useful to include a named function as well.

**2.** rules regarding function calls (Eg. `func-call-spacing`)

**3.** deprecated rules

**Does this change cause the rule to produce more or fewer warnings?**

Same.
## How will the change be implemented?
1. `this program` if global return
2. `${getFunctionNameWithKind(node)}` if function

**array-callback-return**

Before

> Expected to return a value at the end of this function.

> Expected to return a value in this function.

> Expected a return value.

After

> Expected to return a value at the end of `${getFunctionNameWithKind(node)}`.

> Expected to return a value in `${getFunctionNameWithKind(node)}`.

> `${getFunctionNameWithKind(node)}` expected a return value.

**complexity**

Before

> Function '{{name}}' has a complexity of {{complexity}}.

After

> `${getFunctionNameWithKind(node)}` has a complexity of {{complexity}}.

**consistent-return**

Before

> Expected {{which}} return value.

> Expected to return a value at the end of this {{type}}.

After

> `${getFunctionNameWithKind(node)}` expected {{which}} return value.

> Expected to return a value at the end of `${getFunctionNameWithKind(node)}`.

**func-names**

Before

> Unexpected function expression name.

After

> Unexpected named `${getFunctionNameWithKind(node)}`.

**max-params**

Before

> This function has too many parameters ({{count}}). Maximum allowed is {{max}}.

After

> `${getFunctionNameWithKind(node)}` has too many parameters ({{count}}). Maximum allowed is {{max}}.

**max-statements**

Before

> This function has too many statements ({{count}}). Maximum allowed is {{max}}.

After

> `${getFunctionNameWithKind(node)}` has too many statements ({{count}}). Maximum allowed is {{max}}.

**no-empty-functions**

Before

> Unexpected empty {{kind}}.

After

> Unexpected empty `${getFunctionNameWithKind(node)}`.

**strict**

Before

> Wrap this function in a function with 'use strict' directive.

After

> Wrap `${getFunctionNameWithKind(node)}` in a function with 'use strict' directive.

### CHANGELOG
* 30/12
  * Added `${getFunctionNameWithKind(node)}`.
  * Rewrote **func-names** feedback message.

**Comment by platinumazure** [comment]
I like the idea. That said, since we are also discussing a rule change elsewhere about using a function's inferred name to suppress warnings, I would like to make sure we consolidate all that logic in one place (ast-utils maybe?).

**Comment by vitorbal** [comment]
extracting that logic into ast-utils sounds good to me. Besides that, 👍 from me so we just need a champion now.

**Comment by mysticatea** [comment]
```js
if (node.parent.type === "MethodDefinition" || node.parent.type === "Property") {
    name = node.parent.key.name;
}
```

This needs consideration for computed properties (it can become `undefined` on computed properties).
Maybe `name = astUtils.getStaticPropertyName(node.parent) || "(computed)"`?

**Comment by platinumazure** [comment]
I've decided to champion this. Just one more endorsement and we can accept this issue!

**Comment by platinumazure** [comment]
@Moeriki This is accepted now :smile:

You had promised us a broader list of rules that could use this improvement once the issue was accepted. Please present that list at your convenience. I look forward to seeing what we can do here! Thanks in advance!

**Comment by Moeriki** [comment]
I noticed `consistent-return` differentiates in its report messages between functions and methods.

`Expected to return a value at the end of this function.` VS `Expected to return a value at the end of this method.`. Do we want to keep this trend?

**Comment by Moeriki** [comment]
For anonymous functions do you prefer 1️⃣ `this function` or 2️⃣ `function 'anonymous'`?

Complexity does the latter right now. I'm more a fan of the former.

**Comment by Moeriki** [comment]
I updated the issue.

1. Include `complexity`, `consistent-return`, `func-names`, `no-empty-functions`, and `strict`
2. Prefer `this function` over `function 'anonymous'`
3. Prefer `this method` / `method {{name}}` when function is object property / class method

**Comment by platinumazure** [comment]
Looks good to me! (I prefer "this function" over "function 'anonymous'" in the unlikely event someone uses "anonymous" as a function name.)

**Comment by platinumazure** [comment]
Okay, here's where I think we are on this issue:

1. The issue is accepted, so work can begin anytime.
1. We have a good list of rules that should have this logic applied, and PRs can be sent to fix one or many of those at a time.
1. We have centralized logic for getting a function name and kind in a good way (see ast-utils' `getFunctionNameWithKind` method). However, it doesn't use "this" for anonymous functions. Should we augment `getFunctionNameWithKind` to support a parameter for prepending "this" to the names returned for anonymous functions? Or is that not important?

**Comment by platinumazure** [comment]
@Moeriki @mysticatea Do either of you have any thoughts on [my most recent comment](https://github.com/eslint/eslint/issues/7260#issuecomment-263653956)?

**Comment by nzakas** [comment]
For point 3, I'd say open a separate issue. 

For this issue, I think we either need someone to commit to doing it or we should close it. This has been open for three months with no progress, so it's time to make a call.

**Comment by Moeriki** [comment]
I don't know about the need for moving the `this` anonymous function logic to ast-utils.

I'll commit to implementing this.

**Comment by Moeriki** [comment]
[A WIP](https://github.com/eslint/eslint/compare/master...Moeriki:include-function-name-in-report-message).

* I didn't add many tests as astUtils itself is well tested. Is that ok, or do you prefer all cases to be covered per-rule as well?

* Can we drop `this function` in favour of just `function` in the feedback message? Eg. `Function has a complexity of {{complexity}}.` instead of `This function has a complexity of {{complexity}}`. If not, how do we handle the logic on when to prepend _this_ to the name-with-kind?

* I capitalized the first letter of the output of `getFunctionNameWithKind` when the parameter is first part of the feedback message.

* For the `getFunctionNameWithKind` output to be usable in `func-names` I rewrote the feedback message to `Unexpected named {{name}}` / `Unexpected unnamed {{name}}`. Your thoughts on that?

* There is still a whole chunk of logic in `consistent-return` to determine `loc` that is not in the other rules. So location accuracy may be off in other rules.

* One test for `max-statements` failed becaused the expected message included the quotes in the literal. `Method ''thing'' has too many statements (3). Maximum allowed is 2.`. When using `getFunctionNameWithKind`. These quotes are not preserved (`Method 'thing' has…`). Do we mind? If we do mind, ast-utils will have to be updated probably.

* I had to leave the `getKind` function in `no-empty-function` because it's needed to map against the allowed options.

**Comment by platinumazure** [comment]
@Moeriki I'm really sorry for letting this slide. Here are some answers to your questions:

> * I didn't add many tests as astUtils itself is well tested. Is that ok, or do you prefer all cases to be covered per-rule as well?

That should be fine. IMO, rule tests are more about functionality than error messages, and if (say) an async function test is irrelevant to the rule, we don't need to worry about adding one just to test the error message.

> * Can we drop `this function` in favour of just `function` in the feedback message? Eg. `Function has a complexity of {{complexity}}.` instead of `This function has a complexity of {{complexity}}`. If not, how do we handle the logic on when to prepend _this_ to the name-with-kind?

Could you provide a concrete example? I'm not sure I follow.

> * I capitalized the first letter of the output of `getFunctionNameWithKind` when the parameter is first part of the feedback message.

:+1:

> * For the `getFunctionNameWithKind` output to be usable in `func-names` I rewrote the feedback message to `Unexpected named {{name}}` / `Unexpected unnamed {{name}}`. Your thoughts on that?

I think that's probably fine, but at the same time I wouldn't mind if others took a look and provided some feedback as well.

> * There is still a whole chunk of logic in `consistent-return` to determine `loc` that is not in the other rules. So location accuracy may be off in other rules.

Sure, but that's probably a separate issue.

> * One test for `max-statements` failed becaused the expected message included the quotes in the literal. `Method ''thing'' has too many statements (3). Maximum allowed is 2.`. When using `getFunctionNameWithKind`. These quotes are not preserved (`Method 'thing' has…`). Do we mind? If we do mind, ast-utils will have to be updated probably.

Hmm. I think we can just deal with the quote change for now (i.e., don't modify ast-utils), but it would be nice to augment ast-utils with an ability to support particular string formatting options later on.

> * I had to leave the `getKind` function in `no-empty-function` because it's needed to map against the allowed options.

So it goes, I guess.

-------

I think you could probably open a PR anytime (though I would recommend rebasing first). Then more team members can review it. Thanks!

**Comment by Moeriki** [comment]
Great. I rebased onto your master.

> Could you provide a concrete example? I'm not sure I follow.

Eg. `consistent-return` had as message `Expected to return a value at the end of this function.` before my change.

But `astUtils. getFunctionNameWithKind ` returns either `function 'foo'` for named functions or `function` for unnamed functions. So by replacing `…end of this function.` in the afore mentioned message with `…end of {name}.` the feedback message no longer has `…end of this function.` for unnamed functions but just `…end of function.`.

I'm not sure how to solve _this_. So for now I just dropped _this_. 🙃 This impacts `consistent-this` and `strict`.

For the sake of progress I'll just open PR. We'll continue with the feedback I receive there.
